src/command/dev/mod.rs

```diff
@@ -63,38 +74,140 @@ pub struct SupergraphOpts {
 
     /// The address the graph router should listen on.
     ///
     /// If you start multiple `rover dev` processes on the same address and port, they will communicate with each other.
     ///
     /// If you start multiple `rover dev` processes with different addresses and ports, they will not communicate with each other.
-    #[arg(long, default_value = "127.0.0.1")]
-    supergraph_address: String,
+    #[arg(long)]
+    supergraph_address: Option<String>,
+
+    /// The path to a router configuration file. If the file is empty, a default configuration will be written to that file.
+    ///
+    /// For information on the format of this file, please see https://www.apollographql.com/docs/router/configuration/overview/#yaml-config-file.
+    #[arg(long = "router-config", conflicts_with_all = ["supergraph_port", "supergraph_address"])]
+    #[serde(skip_serializing)]
+    router_config_path: Option<Utf8PathBuf>,
+
+    #[arg(skip)]
+    #[serde(skip_serializing)]
+    resolved_router_address: AtomicLazyCell<SocketAddr>,
+
+    #[arg(skip)]
+    #[serde(skip_serializing)]
+    resolved_router_config: AtomicLazyCell<Utf8PathBuf>,
+
+    #[arg(skip)]
+    #[serde(skip_serializing)]
+    temp_dir_path: AtomicLazyCell<Utf8PathBuf>,
 }
 
 #[cfg(feature = "composition-js")]
 impl SupergraphOpts {
-    pub fn router_socket_addr(&self) -> RoverResult<SocketAddr> {
-        let socket_candidate = format!("{}:{}", &self.supergraph_address, &self.supergraph_port);
-        Ok(SocketAddr::from_str(&socket_candidate)
-            .with_context(|| format!("{} is not a valid socket address", &socket_candidate))?)
+    /// Get the name of the socket the router should listen on for incoming GraphQL requests
+    fn router_socket_addr(&self) -> RoverResult<SocketAddr> {
+        if let Some(socket_addr) = self.resolved_router_address.borrow() {
+            Ok(*socket_addr)
+        } else {
+            let socket_candidate = if let Some(router_config) = &self.router_config_path {
+                let contents = Fs::read_file(router_config)?;
+                let yaml: serde_yaml::Mapping = serde_yaml::from_str(&contents)
+                    .with_context(|| format!("'{router_config}' is not valid YAML"))?;
+                if let Some(socket_addr) = yaml
+                    .get("supergraph")
+                    .and_then(|s| s.as_mapping())
+                    .and_then(|s| s.get("listen"))
+                    .and_then(|l| l.as_str())
+                {
+                    socket_addr.to_string()
+                } else {
+                    "127.0.0.1:3000".to_string()
+                }
+            } else {
+                format!(
+                    "{}:{}",
+                    &self
+                        .supergraph_address
+                        .clone()
+                        .unwrap_or_else(|| "127.0.0.1".to_string()),
+                    &self.supergraph_port.unwrap_or(3000)
+                )
+            };
```

Comment:
would this read better as a `match`? Still developing a rusty intuition about code readability, but I found myself jumping back and forth between the if/else blocks to make sure I was still in a particular code path

Reply:
yeah and i can probably split it up into some functions too